### PR TITLE
Move Optional support into Retrofit core.

### LINKED ISSUE #2893: Move Java8 converter and adapter into the core
These should be progressive enhancements based on availability rather than requiring explicit registration and a separate dependency.

### PR DESCRIPTION
**Pull request by NightlyNexus**
Closes #2893.

### CODE REVIEW

File: retrofit/src/test/java/retrofit2/OptionalConverterFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+package retrofit2;
```
Comment (swankjesse) [inline retrofit/src/test/java/retrofit2/OptionalConverterFactoryTest.java]: nit: copyright

File: retrofit/src/main/java/retrofit2/OptionalConverterFactory.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package retrofit2;
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit2/OptionalConverterFactory.java]: nit: copyright

File: retrofit/src/main/java/retrofit2/OptionalConverterFactory.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+ * A {@linkplain Converter.Factory converter} for {@code Optional<T>} which delegates to another
+ * converter to deserialize {@code T} and then wraps it into {@link Optional}.
+ */
+@IgnoreJRERequirement
+final class OptionalConverterFactory extends Converter.Factory {
+  static final Converter.Factory INSTANCE = new OptionalConverterFactory();
+
+  @Nullable @Override
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit2/OptionalConverterFactory.java]: nit: put `@Nullable` immediately preceding the return type, to hint that we interpret it as a type annotation:
```
public @Nullable Converter<ResponseBody, ?> responseBodyConverter(
      Type type, Annotation[] annotations, Retrofit retrofit) {
```

File: retrofit/src/test/java/retrofit2/OptionalConverterFactoryTest.java (hunk 1 above)
Comment (NightlyNexus) [reply retrofit/src/test/java/retrofit2/OptionalConverterFactoryTest.java]: Whoops. Thanks for these catches.

File: retrofit/src/main/java/retrofit2/OptionalConverterFactory.java (hunk 3 above)
Comment (NightlyNexus) [reply retrofit/src/main/java/retrofit2/OptionalConverterFactory.java]: Agree! Will do.

**Pull request merged**